src/runtime/HalideBuffer.h

```diff
@@ -210,37 +193,57 @@ class Buffer {
                 delete dev_ref_count;
             }
         }
-        buf.dev = 0;
+        buf.device = 0;
         dev_ref_count = nullptr;
     }
 
-    /** A temporary helper function to get the number of dimensions in
-     * a buffer_t. Will disappear when halide_buffer_t is merged. */
-    int buffer_dimensions(const buffer_t &buf) {
-        for (int d = 0; d < 4; d++) {
-            if (buf.extent[d] == 0) {
-                return d;
-            }
+    void free_shape_storage() {
+        if (buf.dim != shape) {
+            delete[] buf.dim;
+            buf.dim = nullptr;
+        }
+    }
+
+    void make_shape_storage() {
+        if (buf.dimensions <= D) {
+            buf.dim = shape;
+        } else {
+            buf.dim = new halide_dimension_t[buf.dimensions];
+        }
+    }
+
+    void copy_shape_from(const halide_buffer_t &other) {
+        make_shape_storage();
```

Comment:
assert(buf.dimensions == other.dimensions) ?

Or at least that other.dimensions >= buf.dimensions, to avoid possible bad memory accesses.